src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs

```diff
@@ -189,10 +193,42 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
             }
         }
 
+        /// <summary>
+        /// Try to process a single message after the circuit was broken, a.k.a entering the half-open state.
+        /// </summary>
+        /// <returns>
+        ///     [true] when the related message handler can again process messages and the message pump can again start receive messages in full; [false] otherwise.
+        /// </returns>
+        public override async Task<bool> TryProcessProcessSingleMessageAsync(MessagePumpCircuitBreakerOptions options)
+        {
+            Logger.LogDebug("Azure Service Bus {EntityType} message pump '{JobId}' on entity path '{EntityPath}' in namespace '{Namespace}' tries to process single message during half-open circuit...", Settings.ServiceBusEntity, JobId, EntityPath, Namespace);
+            
+            ServiceBusReceivedMessage message = null;
+            while (message is null)
+            {
+                message = await _messageReceiver.ReceiveMessageAsync();
+            }
+
+            try
+            {
+                await ProcessMessageAsync(new ProcessMessageEventArgs(message, _messageReceiver, CancellationToken.None));
+                return true;
+            }
+            catch (Exception exception)
```

Comment:
The `ProcessMessageAsync` does not return any value. Also the message handlers do not return any value. The method `ProcessMessageAsync` here, is the same method that is called on an 'received message event', registerd on the Service Bus processor.

Reply:
But the `TryProcess...` method does ?   (That's the method that is being overriden / implemented by implementations, isn't it ? )

I think we should incorporate the return value of that method for deciding whether or not the circuit breaker must open.

Reply:
Additionally, it is currently not clear to me where and how we decide when the circuit-breaker should be triggered.
Is it somewhere configurable ?  Can we say:  if x amount of message fail within a certain period, the circuit-breaker should open ?

Reply:
> That's the method that is being overriden / implemented by implementations, isn't it 

Only in custom message pumps, yes, not in custom message handlers.

> Is it somewhere configurable ? Can we say: if x amount of message fail within a certain period, the circuit-breaker should open ?

As I said in my previous comment: https://github.com/arcus-azure/arcus.messaging/pull/416#discussion_r1334011877, this is not yet implemented yet. You think we require this in the first version of the circuit breaker?
I included the `CircuitBreakerOptions` in the `TryProcess..` method to be able to include those in the next iteration (currently only a single successfully processed message is enough to make the circuit breaker close/open).

>  it is currently not clear to me where and how we decide when the circuit-breaker should be triggered.

It is triggered in the custom message handler with the `IMessagePumpCircuitBreaker.PauseMessageProcessingAsync`. That will pause the message processing. The circuit breaker is then activated and internally, a single message will be tried after the initial recovery period. All the custom message handler has to know about is this interface, all the rest is handled internally.

To configure the circuit breaker, the `PauseMessageProcessingAsync` has a `CircuitBreakerOptions` set where you can specify the wait times (and later the amount of messages that needs to be tried once the circuit breaker is half-open).

Hope this helps.
Let me know if we have to update the documentation. Also note: it is for the consumers not important to know how it works internally, right? That's why the documentation only focusses on the `IMessagePumpCircuitBreaker` that can be injected in their custom message handler.

Reply:
Let me know if you believe that we should also implement the amount of messages in the options, then we can look into that before we go on with this PR. 👍

Reply:
Yes, true.  The message-handler triggers the circuit-breaker by calling the `PauseMessagProcessingAsync` method.  However, I don't think that a message-handler will call this method on the very first time processing a message fails.   I think we should have the ability to keep track of the amount of messages that have failed, and use that information to decide if we're going to call the `PauseMessageProcessing` method.

Going further on this, we could even argue if it should be the responsability of the message-handler to pause the processing.   If the message-handler could give feedback about message-failures that should lead to the circuit-breaker being triggered (by increasing a certain counter for instance), that might also be a viable option.  In that case, the pump itself then decides that it should no longer retrieve messages from the queue / topic based on those counters.

My earlier comment that started this discussion was actually posted because I was -at that time- thinking of this kind of implementation.

Reply:
I think it would be nice to have some additional opinions / pov's here.  Maybe @pim-simons and @gverstraete can share their opinion ?

Reply:
We discussed this offline and came to the conclusion that:

- the Message Handler is the one that knows that the dependency call failed, so the Message Handler must be the one that initiates the circuit breaker.
- The way it is implemented right now is OK.  We can do it by increasing a counter in the MessageHandler and let the pump decide when the circuit-break must be triggered.  However, this brings another set of complexities with it.   

Conclusion is that the approach that is taken in this PR is good and we should continue with this approach.